statsd/statsd.go

```diff
@@ -121,30 +135,100 @@ func (c *Client) watch() {
 
 func (c *Client) append(cmd string) error {
 	c.Lock()
+	defer c.Unlock()
 	c.commands = append(c.commands, cmd)
 	// if we should flush, lets do it
 	if len(c.commands) == c.bufferLength {
 		if err := c.flush(); err != nil {
-			c.Unlock()
 			return err
 		}
 	}
-	c.Unlock()
 	return nil
 }
 
+func joinMaxSize(cmds []string, sep string, maxSize int) ([]string, []int) {
+	var buffer bytes.Buffer
+	buffer.Reset() //just to be sure...
+
+	frames := make([]string, 0)
+	ncmds := make([]int, 0)
+	sepLength := len(sep)
+
+	elem := 0
+	for _, cmd := range cmds {
+		var needed int
+
+		if elem == 0 {
+			needed = len(cmd)
+		} else {
+			needed = sepLength + len(cmd)
+		}
+
+		if buffer.Len()+needed <= maxSize {
+			if elem != 0 {
+				buffer.WriteString(sep)
+			}
+			buffer.WriteString(cmd)
+			elem++
+		} else if elem == 0 {
+			//if a single command is too big, flush it and hope the UDP fragmented packet
+			//doesn't get dropped.
+			frames = append(frames, cmd)
+			ncmds = append(ncmds, 1)
+		} else {
+			frames = append(frames, buffer.String())
+			ncmds = append(ncmds, elem)
+			buffer.Reset()
+			buffer.WriteString(cmd)
+			elem = 1
+		}
+	}
+
+	//add whatever is left!
+	frames = append(frames, buffer.String())
+	ncmds = append(ncmds, elem)
+
+	return frames, ncmds
+}
+
 // flush the commands in the buffer.  Lock must be held by caller.
 func (c *Client) flush() error {
-	data := strings.Join(c.commands, "\n")
-	_, err := c.conn.Write([]byte(data))
+	frames, flushable := joinMaxSize(c.commands, "\n", MaxPayloadSize)
+	var err error
+	cmdsFlushed := 0
+	for i, data := range frames {
+		_, e := c.conn.Write([]byte(data))
+		if e != nil {
+			err = e
+			for j := 0; j < i; j++ {
+				cmdsFlushed += flushable[j]
+			}
+			break
+		}
+	}
+	if err == nil {
+		for _, flushed := range flushable {
+			cmdsFlushed += flushed
+		}
+	}
+
 	// clear the slice with a slice op, doesn't realloc
-	c.commands = c.commands[:0]
+	if cmdsFlushed == len(c.commands) {
+		c.commands = c.commands[:0]
+	} else {
+		//this case would cause an eventual realloc in the future...
+		c.commands = c.commands[cmdsFlushed:]
```

Comment:
What would happen if `commands` is full and we can't send any command on a `flush` call? This would lead to a sort of overflow, commands would grow larger than specified capacity.